File: PE-diagram/src/main/java/nl/adaptivity/process/diagram/DrawableJoin.java
/*
 * Copyright (c) 2016.
 *
 * This file is part of ProcessManager.
 *
 * ProcessManager is free software: you can redistribute it and/or modify it under the terms of version 3 of the
 * GNU Lesser General Public License as published by the Free Software Foundation.
 *
 * ProcessManager is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even
 * the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License along with ProcessManager.  If not,
 * see <http://www.gnu.org/licenses/>.
 */

package nl.adaptivity.process.diagram;

import nl.adaptivity.diagram.*;
import nl.adaptivity.process.clientProcessModel.ClientJoinNode;
import nl.adaptivity.process.processModel.IXmlDefineType;
import nl.adaptivity.process.processModel.IXmlResultType;
import nl.adaptivity.process.processModel.Join;
import nl.adaptivity.process.processModel.ModelCommon;
import nl.adaptivity.process.util.Identified;
import nl.adaptivity.xml.XmlException;
import nl.adaptivity.xml.XmlReader;
import nl.adaptivity.xml.XmlUtil;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.Collection;

import static nl.adaptivity.process.diagram.RootDrawableProcessModel.*;


public class DrawableJoin extends ClientJoinNode<DrawableProcessNode, DrawableProcessModel> implements Join<DrawableProcessNode, DrawableProcessModel>, DrawableJoinSplit {

  public static class Builder extends ClientJoinNode.Builder<DrawableProcessNode, DrawableProcessModel> implements DrawableJoinSplit.Builder {

    public Builder() {
    }

    public Builder(final boolean compat) {
      super(compat);
    }

    public Builder(@NotNull final Collection<? extends Identified> predecessors, @NotNull final Identified successor, @Nullable final String id, @Nullable final String label, final double x, final double y, @NotNull final Collection<? extends IXmlDefineType> defines, @NotNull final Collection<? extends IXmlResultType> results, final int min, final int max) {
      super(predecessors, successor, id, label, x, y, defines, results, min, max);
    }

    public Builder(@NotNull final Join<?, ?> node) {
      super(node);
    }

    @NotNull
    @Override
    public DrawableJoin build(@NotNull final DrawableProcessModel newOwner) {
      return new DrawableJoin(this, newOwner);
    }
  }

  private static final double ARROWHEADDX = JOINWIDTH*0.375;
  private static final double ARROWHEADADJUST = 0.5*STROKEWIDTH/Math.sin(DrawableJoinSplit.ARROWHEADANGLE);

  /** The y coordinate if the line were horizontal. */
  private static final double ARROWDFAR = DrawableJoinSplit.ARROWLEN*Math.sin(DrawableJoinSplit.ARROWHEADANGLE);
  /** The x coordinate if the line were horizontal. */
  private static final double ARROWDNEAR = DrawableJoinSplit.ARROWLEN*Math.cos(DrawableJoinSplit.ARROWHEADANGLE);
  private static final double INDX = JOINWIDTH*0.2;
  private static final double INDY = JOINHEIGHT*0.2;
  private static final double INLEN = Math.sqrt(INDX*INDX+INDY*INDY);

  private final DrawableJoinSplitDelegate mDrawableJoinSplitDelegate;

  public DrawableJoin(final DrawableProcessModel ownerModel) {
    this(ownerModel, false);
  }

  public DrawableJoin(final DrawableProcessModel ownerModel, final boolean compat) {
    super(ownerModel, compat);
    mDrawableJoinSplitDelegate = new DrawableJoinSplitDelegate();
  }

  public DrawableJoin(final DrawableProcessModel ownerModel, String id, final boolean compat) {
    super(ownerModel, id, compat);
    mDrawableJoinSplitDelegate = new DrawableJoinSplitDelegate();
  }

  public DrawableJoin(DrawableJoin orig, final boolean compat) {
    super(orig, compat);
    mDrawableJoinSplitDelegate = new DrawableJoinSplitDelegate(orig.mDrawableJoinSplitDelegate);
  }

  public DrawableJoin(@NotNull final ClientJoinNode.Builder<?, ?> builder, @NotNull final DrawableProcessModel newOwnerModel) {
    super(builder, newOwnerModel);
    mDrawableJoinSplitDelegate = new DrawableJoinSplitDelegate();
  }

  @NotNull
  @Override
  public Builder builder() {
    return new Builder(this);
  }

  @Override
  public DrawableJoin clone() {
    if (getClass()==DrawableJoin.class) {
      return new DrawableJoin(this, isCompat());
    }
    throw new RuntimeException(new CloneNotSupportedException());
  }

  public static DrawableJoin from(Join<?, ?> elem, final boolean compat) {
    return new DrawableJoin((DrawableProcessModel)null, compat);
  }

  @NotNull
  public static DrawableJoin deserialize(final DrawableProcessModel ownerModel, @NotNull final XmlReader in) throws XmlException {
    return XmlUtil.<Builder>deserializeHelper(new Builder(true), in).build(ownerModel);
  }

  @NotNull
  public static Builder deserialize(@NotNull final XmlReader in) throws XmlException {
    return XmlUtil.deserializeHelper(new Builder(true), in);
  }

  @Override
  public <S extends DrawingStrategy<S, PEN_T, PATH_T>, PEN_T extends Pen<PEN_T>, PATH_T extends DiagramPath<PATH_T>> void draw(Canvas<S, PEN_T, PATH_T> canvas, Rectangle clipBounds) {
    mDrawableJoinSplitDelegate.draw(this, canvas, clipBounds);

    final S strategy = canvas.getStrategy();
    PATH_T path = mDrawableJoinSplitDelegate.mItems.getPath(strategy, 1);
    if (path==null) {
      path = strategy.newPath();
      if (CURVED_ARROWS) {
        path.moveTo(CENTERX+INLEN,CENTERY)
            .lineTo(CENTERX + ARROWHEADDX - ARROWHEADADJUST, CENTERX)
            .moveTo(CENTERX+ARROWHEADDX-ARROWDNEAR, CENTERY-ARROWDFAR)
            .lineTo(CENTERX+ARROWHEADDX, CENTERY)
            .lineTo(CENTERX+ARROWHEADDX-ARROWDNEAR, CENTERY+ARROWDFAR)
            .moveTo(CENTERX-INDX,CENTERY-INDY)
            .cubicTo(CENTERX-(INDX*(1-ARROWCONTROLRATIO)), CENTERY-(INDY*(1-ARROWCONTROLRATIO)), CENTERX+INLEN*(1-ARROWCONTROLRATIO), CENTERY, CENTERX+INLEN, CENTERY)
            .cubicTo(CENTERX+INLEN*(1-ARROWCONTROLRATIO), CENTERY, CENTERX-(INDX*(1-ARROWCONTROLRATIO)), CENTERY+(INDY*(1-ARROWCONTROLRATIO)), CENTERX-INDX,CENTERY+INDY);
      } else {
        path.moveTo(CENTERX,CENTERY)
            .lineTo(CENTERX + ARROWHEADDX - ARROWHEADADJUST, CENTERX)
            .moveTo(CENTERX+ARROWHEADDX-ARROWDNEAR, CENTERY-ARROWDFAR)
            .lineTo(CENTERX+ARROWHEADDX, CENTERY)
            .lineTo(CENTERX+ARROWHEADDX-ARROWDNEAR, CENTERY+ARROWDFAR)
            .moveTo(CENTERX-INDX,CENTERY-INDY)
            .lineTo(CENTERX, CENTERY)
            .lineTo(CENTERX-INDX,CENTERY+INDY);
      }
      mDrawableJoinSplitDelegate.mItems.setPath(strategy, 1, path);
    }
    if (hasPos()) {
      PEN_T linePen = canvas.getTheme().getPen(ProcessThemeItems.INNERLINE, getState());
      canvas.drawPath(path, linePen, null);
    }
  }

  @Override
  public <S extends DrawingStrategy<S, PEN_T, PATH_T>, PEN_T extends Pen<PEN_T>, PATH_T extends DiagramPath<PATH_T>> void drawLabel(Canvas<S, PEN_T, PATH_T> canvas, Rectangle clipBounds, double left, double top) {
    DrawableUtil.defaultDrawLabel(this, canvas, clipBounds, left, top);
  }

  @Override
  public void setPos(final double left, final double top) {
    setX(left);
    setY(top);
  }

  @Override
  public Rectangle getBounds() {
    return DrawableJoinSplitDelegate.getBounds(this);
  }

  @Override
  public Drawable getItemAt(double x, double y) {
    return DrawableJoinSplitDelegate.getItemAt(this, x, y);
  }

  @Override
  public int getState() {
    return mDrawableJoinSplitDelegate.getState();
  }

  @Override
  public void setState(int state) {
    mDrawableJoinSplitDelegate.setState(state);
  }

  public String getMinMaxText() {
    return DrawableJoinSplitDelegate.getMinMaxText(this);
  }

  @Override
  public boolean isOr() {
    return DrawableJoinSplitDelegate.isOr(this);
  }

  @Override
  public boolean isXor() {
    return DrawableJoinSplitDelegate.isXor(this);
  }

  @Override
  public boolean isAnd() {
    return DrawableJoinSplitDelegate.isAnd(this);
  }
}

File: PE-diagram/src/commonMain/kotlin/nl.adaptivity.process/diagram/DrawableJoin.kt
/*
 * Copyright (c) 2017.
 *
 * This file is part of ProcessManager.
 *
 * ProcessManager is free software: you can redistribute it and/or modify it under the terms of version 3 of the
 * GNU Lesser General Public License as published by the Free Software Foundation.
 *
 * ProcessManager is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even
 * the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License along with ProcessManager.  If not,
 * see <http://www.gnu.org/licenses/>.
 */

package nl.adaptivity.process.diagram

import nl.adaptivity.diagram.*
import nl.adaptivity.process.diagram.DrawableJoinSplit.Companion.ARROWCONTROLRATIO
import nl.adaptivity.process.diagram.DrawableJoinSplit.Companion.CENTER_X
import nl.adaptivity.process.diagram.DrawableJoinSplit.Companion.CENTER_Y
import nl.adaptivity.process.diagram.DrawableJoinSplit.Companion.CURVED_ARROWS
import nl.adaptivity.process.diagram.RootDrawableProcessModel.Companion.JOINHEIGHT
import nl.adaptivity.process.diagram.RootDrawableProcessModel.Companion.JOINWIDTH
import nl.adaptivity.process.diagram.RootDrawableProcessModel.Companion.STROKEWIDTH
import nl.adaptivity.process.processModel.*
import nl.adaptivity.process.util.Identified
import kotlin.jvm.JvmStatic
import kotlin.math.cos
import kotlin.math.sin
import kotlin.math.sqrt

interface IDrawableJoin : IDrawableJoinSplit {

    override val maxPredecessorCount: Int get() = Int.MAX_VALUE

    override fun <S : DrawingStrategy<S, PEN_T, PATH_T>, PEN_T : Pen<PEN_T>, PATH_T : DiagramPath<PATH_T>> drawDecoration(
        canvas: Canvas<S, PEN_T, PATH_T>,
        clipBounds: Rectangle?) {
        if (hasPos()) {
            val path = itemCache.getPath(canvas.strategy, 1) {
                if (CURVED_ARROWS) {
                    moveTo(CENTER_X + INLEN, CENTER_Y)
                    lineTo(CENTER_X + ARROWHEADD_X - ARROWHEAD_ADJUST, CENTER_Y)
                    moveTo(CENTER_X + ARROWHEADD_X - ARROWDNEAR, CENTER_Y - ARROWDFAR)
                    lineTo(CENTER_X + ARROWHEADD_X, CENTER_Y)
                    lineTo(CENTER_X + ARROWHEADD_X - ARROWDNEAR, CENTER_Y + ARROWDFAR)
                    moveTo(CENTER_X - IND_X, CENTER_Y - IND_Y)
                    cubicTo(CENTER_X - IND_X * (1 - ARROWCONTROLRATIO), CENTER_Y - IND_Y * (1 - ARROWCONTROLRATIO),
                            CENTER_X + INLEN * (1 - ARROWCONTROLRATIO), CENTER_Y, CENTER_X + INLEN, CENTER_Y)
                    cubicTo(CENTER_X + INLEN * (1 - ARROWCONTROLRATIO), CENTER_Y,
                            CENTER_X - IND_X * (1 - ARROWCONTROLRATIO), CENTER_Y + IND_Y * (1 - ARROWCONTROLRATIO),
                            CENTER_X - IND_X, CENTER_Y + IND_Y)
                    moveTo(CENTER_X + INLEN, CENTER_Y)
                    lineTo(CENTER_X + ARROWHEADD_X, CENTER_Y)
                } else {
                    moveTo(CENTER_X, CENTER_Y)
                    lineTo(CENTER_X + ARROWHEADD_X - ARROWHEAD_ADJUST, CENTER_X)
                    moveTo(CENTER_X + ARROWHEADD_X - ARROWDNEAR, CENTER_Y - ARROWDFAR)
                    lineTo(CENTER_X + ARROWHEADD_X, CENTER_Y)
                    lineTo(CENTER_X + ARROWHEADD_X - ARROWDNEAR, CENTER_Y + ARROWDFAR)
                    moveTo(CENTER_X - IND_X, CENTER_Y - IND_Y)
                    lineTo(CENTER_X, CENTER_Y)
                    lineTo(CENTER_X - IND_X, CENTER_Y + IND_Y)
                }
            }

            val linePen = canvas.theme.getPen(ProcessThemeItems.INNERLINE, state)
            canvas.drawPath(path, linePen, null)
        }
    }

    companion object {
        private const val ARROWHEADD_X = JOINWIDTH * 0.375
        private val ARROWHEAD_ADJUST = 0.5 * STROKEWIDTH / sin(DrawableJoinSplit.ARROWHEADANGLE)

        /** The y coordinate if the line were horizontal.  */
        private val ARROWDFAR = DrawableJoinSplit.ARROWLEN * sin(DrawableJoinSplit.ARROWHEADANGLE)
        /** The x coordinate if the line were horizontal.  */
        private val ARROWDNEAR = DrawableJoinSplit.ARROWLEN * cos(DrawableJoinSplit.ARROWHEADANGLE)
        private const val IND_X = JOINWIDTH * 0.2
        private const val IND_Y = JOINHEIGHT * 0.2
        private val INLEN = sqrt(IND_X * IND_X + IND_Y * IND_Y)
    }

}

class DrawableJoin(builder: Join.Builder,
                   buildHelper: ProcessModel.BuildHelper<*, *, *, *>) : JoinBase<DrawableProcessNode, DrawableProcessModel?>(
    builder, buildHelper), Join, DrawableJoinSplit {

    class Builder : JoinBase.Builder, DrawableJoinSplit.Builder<DrawableJoin>, IDrawableJoin {

        override val _delegate: DrawableProcessNode.Builder.Delegate

        constructor() : this(id = null)

        constructor(id: String? = null,
                    predecessors: Collection<Identified> = emptyList(),
                    successor: Identified? = null,
                    label: String? = null,
                    defines: Collection<IXmlDefineType> = emptyList(),
                    results: Collection<IXmlResultType> = emptyList(),
                    x: Double = Double.NaN,
                    y: Double = Double.NaN,
                    min: Int = 1,
                    max: Int = -1,
                    state: DrawableState = Drawable.STATE_DEFAULT,
                    isMultiMerge: Boolean = false,
                    isMultiInstance: Boolean = false,
                    isCompat: Boolean = false) : super(id, predecessors, successor, label, defines, results,
                                                       x, y, min, max, isMultiMerge, isMultiInstance) {
            _delegate = DrawableProcessNode.Builder.Delegate(state, isCompat)
        }

        constructor(node: Join) : super(node) {
            _delegate = DrawableProcessNode.Builder.Delegate(node)
        }

        override val itemCache = ItemCache()

        override fun copy(): Builder =
            Builder(id, predecessors, successor?.identifier, label, defines, results, x, y, min, max, state,
                    isMultiMerge, isMultiInstance, isCompat)
    }

    override val _delegate: DrawableJoinSplit.Delegate

    override val maxSuccessorCount: Int
        get() = if (isCompat) Int.MAX_VALUE else 1

    override val maxPredecessorCount get() = super<JoinBase>.maxPredecessorCount

    override fun builder(): Builder {
        return Builder(this)
    }

    companion object {

        private const val ARROWHEADD_X = JOINWIDTH * 0.375
        private val ARROWHEAD_ADJUST = 0.5 * STROKEWIDTH / sin(DrawableJoinSplit.ARROWHEADANGLE)

        /** The y coordinate if the line were horizontal.  */
        private val ARROWDFAR = DrawableJoinSplit.ARROWLEN * sin(DrawableJoinSplit.ARROWHEADANGLE)
        /** The x coordinate if the line were horizontal.  */
        private val ARROWDNEAR = DrawableJoinSplit.ARROWLEN * cos(DrawableJoinSplit.ARROWHEADANGLE)
        private const val IND_X = JOINWIDTH * 0.2
        private const val IND_Y = JOINHEIGHT * 0.2
        private val INLEN = sqrt(IND_X * IND_X + IND_Y * IND_Y)

        @Deprecated("Use the builder",
                    ReplaceWith("Builder(elem).build()", "nl.adaptivity.process.diagram.DrawableJoin.Builder"))
        @JvmStatic
        fun from(elem: Join, compat: Boolean): DrawableJoin {
            return Builder(elem).build()
        }

    }

    init {
        _delegate = DrawableJoinSplit.Delegate(builder)
    }
}
